sympy/stats/matrix_distributions.py

```diff
@@ -77,15 +77,15 @@ def check(alpha, beta, scale_matrix):
         if not isinstance(scale_matrix , MatrixSymbol):
             _value_check(scale_matrix.is_positive_definite, "The shape "
                 "matrix must be positive definite.")
-        _value_check(scale_matrix.shape[0] == scale_matrix.shape[1], "Should "
+        _value_check(scale_matrix.is_square, "Should "
         "be square matrix")
         _value_check(alpha.is_positive, "Shape parameter should be positive.")
         _value_check(beta.is_positive, "Scale parameter should be positive.")
 
     @property
     def set(self):
         k = self.scale_matrix.shape[0]
-        return S.Reals ** k
+        return ProductSet(S.Reals**k, S.Reals**k)
```

Comment:
Are you sure this is the correct way to call `ProductSet`?

Reply:
I don't know this code but that would be the correct way if you wanted to create a set of 2-tuples where each element is a k-tuple e.g.:
```
In [8]: S = ProductSet(Reals**3, Reals**3)                                                                             

In [9]: ((1, 2, 3), (4, 5, 6)) in S                                                                                    
Out[9]: True
```

Reply:
I meant to represent <b>R<sup>k x k</sup></b> using the `ProductSet`

Reply:
Translating that literally would give `Reals ** (k * k)` which is the set of k-tuples of k-tuples of real numbers. If you mean for this to be the set of `k x k` *matrices* over the reals then you'll need a new set class to represent that.

Reply:
>  If you mean for this to be the set of `k x k` _matrices_ over the reals

Yes, I meant to represent `k x k` _matrices_ over the reals.

Reply:
Should the new class be added under `sets/fancysets.py`?
Also is the following code is going correct way as I'm not much familiar with `sets` code?
```py
class MatrixSet(Set):
    is_iterable = False
    is_empty = False
    is_finite_set = False

    def __new__(cls, shape=(), set=Reals):
        shape = (_sympify(x) for x in dimensions)
        set = _sympify(set)
        return Set.__new__(cls, shape, set)

    @property
    def shape(self):
        return self.args[0]

    @property
    def set(self):
        return self.args[1]

    def _contains(self, other):
        from sympy.matrices import ImmutableMatrix, MatrixExpr
        if isinstance(other, list):
            other = ImmutableMatrix(other)
        if not isinstance(other, MatrixExpr):
            return False
        if other.shape != self.shape:
            return False
        # check that each element of other is in self.set

    def _eval_is_subset(self, other):
        pass

    def _eval_is_superset(self, other):
        pass

```

Reply:
I think I wold put this in the matrices module rather than the sets module because the sets module is lower in the import chain.

The Set class looks reasonable except:

1. I don't see why any of the arguments to `__new__` should be optional.

2. Don't try to handle things like lists. The user should be using a MatrixExpr or something that would sympify to a MatrixExpr (e.g. mutable Matrix) and it's better for them to see a TypeError if they pass something else.

3. You can check whether the elements are in set with:
```
return fuzzy_and(self.set.contains(x) for x in other)
```

4. The set can be finite if the underlying set is finite.

5. There's no need for the `_eval_is_subset` methods.

Reply:
> I think I wold put this in the matrices module rather than the sets module because the sets module is lower in the import chain.

I'm afraid that it can end up with sets sprayed over different places for `Matrix, Vector, Array, ...` and I rarely expect people to put the category of MatrixSet in `matrices` than `sets`.
Actually, MatrixSet would be on top of import chains of both `matrices` and `sets` module, and the cyclic import problem may be avoided by just creating a new file.

Reply:
Personally I would expect to find sets of matrices in the matrices package just like I would expect them to be defined in the wikipedia page for matrices rather than for sets or in a textbook on matrices rather than sets etc.

It's not just the cyclic import problem but the logical organisation of the code. The sets module should be a more fundamental module than matrices. If we insist on defining all sets there then it will end up importing every other package. The MatrixSet class will have almost no interaction with the other sets.

Reply:
I don’t think that the matrix set can be decoupled from set module.

For example, once we start to support new types of objects for sets, FiniteSet, ImageSet, ConditionSet sets should be updated for them, because we will arrive into a question like ‘Why FiniteSet of matrices is not working with MatrixSet‘. And it can likely propagate up to every composite set objects.

Reply:
FiniteSet should not need to know anything about its elements. If we have to add special cases to those classes for matrices then something is wrong in the general structure.

Reply:
I also expect sets to work only relying on the knowledge of containment and subset relation, but I think that there should be lots of simplification techniques that can make use of the nature of its elements.